// node-upload-util/dist/sftp_uploader.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "ssh2-sftp-client", "fs", "path", "./base_uploader", "./util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.SftpUploader = void 0;
    const tslib_1 = require("tslib");
    const Client = require("ssh2-sftp-client");
    const fs = require("fs");
    const path = require("path");
    const base_uploader_1 = require("./base_uploader");
    const util_1 = require("./util");
    class SftpUploader extends base_uploader_1.BaseUploader {
        initOptions(options) {
            this.options = options;
        }
        connect() {
            return tslib_1.__awaiter(this, void 0, void 0, function* () {
                try {
                    this.client = new Client();
                    yield this.client.connect({
                        host: this.options.host,
                        port: this.options.port,
                        username: this.options.user,
                        password: this.options.password
                    });
                    this.onReady();
                }
                catch (e) {
                    throw new Error(e);
                }
            });
        }
        upload(filePath, destPath) {
            if (fs.statSync(filePath).isDirectory()) {
                return this.client.mkdir(destPath, true);
            }
            return this.client.put(filePath, destPath);
        }
        startUpload() {
            return tslib_1.__awaiter(this, void 0, void 0, function* () {
                try {
                    const parsedFiles = util_1.parseFiles(this.options.files);
                    const getRealPath = (filePath) => path.join(this.options.rootPath || process.cwd(), filePath);
                    const getDestPath = (filePath) => path.posix.join(this.options.destRootPath, filePath);
                    this.onStart(parsedFiles);
                    for (const filePath of parsedFiles) {
                        yield this.upload(getRealPath(filePath), getDestPath(filePath));
                        this.onFileUpload(filePath, parsedFiles);
                    }
                    yield this.client.end();
                    this.onSuccess(parsedFiles);
                }
                catch (error) {
                    yield this.client.end();
                    this.onFailure(error);
                }
            });
        }
        onDestoryed() {
            if (this.client) {
                this.client = null;
            }
            super.onDestoryed();
        }
    }
    exports.SftpUploader = SftpUploader;
});
